src/dlpack.rs

```diff
@@ -0,0 +1,94 @@
+use core::ptr::NonNull;
+use std::marker::PhantomData;
+
+use dlpark::prelude::*;
+
+use crate::{ArrayBase, Dimension, IntoDimension, IxDyn, ManagedArray, RawData};
+
+impl<A, S, D> ToTensor for ArrayBase<S, D>
+where
+    A: InferDtype,
+    S: RawData<Elem = A>,
+    D: Dimension,
+{
+    fn data_ptr(&self) -> *mut std::ffi::c_void {
+        self.as_ptr() as *mut std::ffi::c_void
+    }
+
+    fn byte_offset(&self) -> u64 {
+        0
+    }
+
+    fn device(&self) -> Device {
+        Device::CPU
+    }
+
+    fn dtype(&self) -> DataType {
+        A::infer_dtype()
+    }
+
+    fn shape(&self) -> CowIntArray {
+        dlpark::prelude::CowIntArray::from_owned(
+            self.shape().into_iter().map(|&x| x as i64).collect(),
+        )
+    }
+
+    fn strides(&self) -> Option<CowIntArray> {
+        Some(dlpark::prelude::CowIntArray::from_owned(
+            self.strides().into_iter().map(|&x| x as i64).collect(),
+        ))
+    }
+}
+
+pub struct ManagedRepr<A> {
+    managed_tensor: ManagedTensor,
+    _ty: PhantomData<A>,
+}
+
+impl<A> ManagedRepr<A> {
+    pub fn new(managed_tensor: ManagedTensor) -> Self {
+        Self {
+            managed_tensor,
+            _ty: PhantomData,
+        }
+    }
+
+    pub fn as_slice(&self) -> &[A] {
+        self.managed_tensor.as_slice()
+    }
+
+    pub fn as_ptr(&self) -> *const A {
+        self.managed_tensor.data_ptr() as *const A
+    }
+}
+
+unsafe impl<A> Sync for ManagedRepr<A> where A: Sync {}
+unsafe impl<A> Send for ManagedRepr<A> where A: Send {}
+
+impl<A> FromDLPack for ManagedArray<A, IxDyn> {
+    fn from_dlpack(dlpack: NonNull<dlpark::ffi::DLManagedTensor>) -> Self {
```

Comment:
this function takes a raw pointer (wrapped in NonNull) and it must be an unsafe function, otherwise we can trivially violate memory safety unfortunately.

The only way to remove this requirement would be if you have a "magical" function that can take an arbitrary pointer and say whether it's a valid, live, non-mutably aliased pointer to a tensor.

Here's how to create a dangling bad pointer:  `NonNull::new(1 as *mut u8 as *mut dlpark::ffi::DLManagedTensor)`  does this code crash if we run with this pointer? I think it would..

Reply:
I agree with you. `from_dlpack` should be unsafe, and users should use it at their own risk.

Reply:
I would say, we normally don't commit to public dependencies that are not stable (yes, not a great policy since ndarray itself is not so stable.), and dlpark is a public dependency here because it becomes part of our API. It could mean it takes a long time between version bumps.